### PR TITLE
[sailfish-webview] Add folderpickerpage to DownloadMenuItem. Contributes to JB#51673

### PR DESCRIPTION
**Pull request by atatarov**

### CODE REVIEW

File: import/popups/DownloadMenuItem.qml [hunk 1]
```diff
@@ -36,7 +37,7 @@ MenuItem {
         return DownloadHelper.createUniqueFileUrl(leafName, targetDirectory)
     }
 
-    onClicked: {
+    function startDownload() {
```
Comment (llewelld) [inline import/popups/DownloadMenuItem.qml:40]: It looks to me like you'd now be better off dropping `downloadFileTargetUrl` entirely as a property and evaluating it here instead. You'd need to remove it from ContextMenu.qml too (which I think is also the correct thing to do).

File: import/popups/DownloadMenuItem.qml [hunk 2]
```diff
@@ -12,6 +12,7 @@
 import QtQuick 2.0
 import Sailfish.Silica 1.0
 import Sailfish.WebEngine 1.0
+import Sailfish.Pickers 1.0
 
 MenuItem {
     id: root
```
Comment (llewelld) [inline import/popups/DownloadMenuItem.qml:18]: The translation hints in ContextMenu.qml for "Save link" and "Save image" no longer apply, so should probably be updated with this PR too:
```
//: This menu item saves link to downloads.
//: This menu item saves image to Gallery application
```

File: import/popups/DownloadMenuItem.qml [hunk 3]
```diff
@@ -48,4 +49,28 @@ MenuItem {
...
+            startDownload()
+        } else {
+            pageStack.animatorPush(folderPickerPage)
+        }
+    }
+
+    Component {
+        id: folderPickerPage
+
+        FolderPickerPage {
```
Comment (llewelld) [inline import/popups/DownloadMenuItem.qml:65]: Can the folder picker be set to start in the most appropriate place (`StandardPaths.download`, or the last selected folder as appropriate)?

File: import/popups/DownloadMenuItem.qml [hunk 4]
```diff
@@ -12,6 +12,7 @@
 import QtQuick 2.0
 import Sailfish.Silica 1.0
 import Sailfish.WebEngine 1.0
+import Sailfish.Pickers 1.0
```
Comment (llewelld) [inline import/popups/DownloadMenuItem.qml:15]: Copyright header needs updating.

File: import/popups/DownloadMenuItem.qml (hunk 3 above)
Comment (atatarov) [reply import/popups/DownloadMenuItem.qml:63]: There is a problem, when you initialize rootPath for FolderPickerDIalog, you can't go to back. It limits a user  to save any directory above last selected folder

Comment (llewelld) [reply import/popups/DownloadMenuItem.qml:63]: Yes, I recall this being a limitation of FolderPicker. I think this is something that will need to be addressed, as it's part of the design: 'By default "Downloads" directory should be shown, but when user chooses another directory and saves the file to it, **next time browser should offer this directory first**' (my emphasis).

What would you propose? We can fix FolderPicker first then return to this, or merge this, fix FolderPicker, then create a new PR to make changes here?

**Comment by llewelld** [comment]
One other things that occurred: would it make sense to expose `downloadsEnabled` in WebView.qml? Currently it's [set to `false`](https://github.com/sailfishos/sailfish-components-webview/blob/fef09bee038016e8396a3241e53283e07a1207f8/import/webview/WebView.qml#L161) and can't be changed by application developers.

**Comment by llewelld** [comment]
After the changes to the supporting components, I think this PR may now need updating.

Starting a webview on a clean install I get the following error when attempting a download:
```
JSComp: HelperAppDialog.js loaded
HelperAppLauncherDialog show
JavaScript error: file:///usr/lib/mozembedlite/components/HelperAppDialog.js, line 84: NS_ERROR_UNEXPECTED: Component returned failure code: 0x8000ffff (NS_ERROR_UNEXPECTED) [nsIPrefBranch.getStringPref]
CONSOLE message:
[JavaScript Error: "NS_ERROR_UNEXPECTED: Component returned failure code: 0x8000ffff (NS_ERROR_UNEXPECTED) [nsIPrefBranch.getStringPref]" {file: "file:///usr/lib/mozembedlite/components/HelperAppDialog.js" line: 84}]
```
If I then add the two `browser.download.` preferences manually to get around this, I then get the following error:
```
[JavaScript Error: "[Exception... "Component returned failure code: 0x80004005 (NS_ERROR_FAILURE) [nsIObserverService.removeObserver]"  nsresult: "0x80004005 (NS_ERROR_FAILURE)"  location: "JS frame :: file:///usr/lib/mozembedlite/components/HelperAppDialog.js :: saveAndDownload/< :: line 159"  data: no]"]
saveAndDownload/<@file:///usr/lib/mozembedlite/components/HelperAppDialog.js:159:7
saveAndDownload@file:///usr/lib/mozembedlite/components/HelperAppDialog.js:140:6
hald_promptForSaveToFileAsync@file:///usr/lib/mozembedlite/components/HelperAppDialog.js:94:9
hald_show@file:///usr/lib/mozembedlite/components/HelperAppDialog.js:76:5
```

**Comment by llewelld** [comment]
Some clarification on my previous comment. If I use [this example webview code](https://github.com/black-sheep-dev/harbour-webview-example), on a clean install (i.e. after deleting `~/.cache/harbour-webview` where the `.mozilla` folder can be found), I get the following error when attempting to download a file:
```
JavaScript error: file:///usr/lib/mozembedlite/components/HelperAppDialog.js, line 84: NS_ERROR_UNEXPECTED: Component returned failure code: 0x8000ffff (NS_ERROR_UNEXPECTED) [nsIPrefBranch.getStringPref]
```
If I then add the preferences in like this:
```
echo 'user_pref("browser.download.useDownloadDir", true);' >> ~/.cache/harbour-webview/harbour-webview/.mozilla/prefs.js
echo 'user_pref("browser.download.dir", "/home/defaultuser/Downloads");' >> ~/.cache/harbour-webview/harbour-webview/.mozilla/prefs.js
```
run  the app and try downloading a file I then get the following error:
```
[JavaScript Error: "[Exception... "Component returned failure code: 0x80004005 (NS_ERROR_FAILURE) [nsIObserverService.removeObserver]"  nsresult: "0x80004005 (NS_ERROR_FAILURE)"  location: "JS frame :: file:///usr/lib/mozembedlite/components/HelperAppDialog.js :: saveAndDownload/< :: line 159"  data: no]"]
saveAndDownload/<@file:///usr/lib/mozembedlite/components/HelperAppDialog.js:159:7
saveAndDownload@file:///usr/lib/mozembedlite/components/HelperAppDialog.js:140:6
hald_promptForSaveToFileAsync@file:///usr/lib/mozembedlite/components/HelperAppDialog.js:94:9
hald_show@file:///usr/lib/mozembedlite/components/HelperAppDialog.js:76:5
```
If I set "browser.download.useDownloadDir" to false, there is no longer any error, but the file doesn't download fully (I get a `.part` file but not the final download file).

Are you able to reproduce this? It could be an issue with my set up here.

**Comment by atatarov** [comment]
> Some clarification on my previous comment. If I use [this example webview code](https://github.com/black-sheep-dev/harbour-webview-example), on a clean install (i.e. after deleting `~/.cache/harbour-webview` where the `.mozilla` folder can be found), I get the following error when attempting to download a file:
> 
> ```
> JavaScript error: file:///usr/lib/mozembedlite/components/HelperAppDialog.js, line 84: NS_ERROR_UNEXPECTED: Component returned failure code: 0x8000ffff (NS_ERROR_UNEXPECTED) [nsIPrefBranch.getStringPref]
> ```
> 
> If I then add the preferences in like this:
> 
> ```
> echo 'user_pref("browser.download.useDownloadDir", true);' >> ~/.cache/harbour-webview/harbour-webview/.mozilla/prefs.js
> echo 'user_pref("browser.download.dir", "/home/defaultuser/Downloads");' >> ~/.cache/harbour-webview/harbour-webview/.mozilla/prefs.js
> ```
> 
> run the app and try downloading a file I then get the following error:
> 
> ```
> [JavaScript Error: "[Exception... "Component returned failure code: 0x80004005 (NS_ERROR_FAILURE) [nsIObserverService.removeObserver]"  nsresult: "0x80004005 (NS_ERROR_FAILURE)"  location: "JS frame :: file:///usr/lib/mozembedlite/components/HelperAppDialog.js :: saveAndDownload/< :: line 159"  data: no]"]
> saveAndDownload/<@file:///usr/lib/mozembedlite/components/HelperAppDialog.js:159:7
> saveAndDownload@file:///usr/lib/mozembedlite/components/HelperAppDialog.js:140:6
> hald_promptForSaveToFileAsync@file:///usr/lib/mozembedlite/components/HelperAppDialog.js:94:9
> hald_show@file:///usr/lib/mozembedlite/components/HelperAppDialog.js:76:5
> ```
> 
> If I set "browser.download.useDownloadDir" to false, there is no longer any error, but the file doesn't download fully (I get a `.part` file but not the final download file).
> 
> Are you able to reproduce this? It could be an issue with my set up here.

My steps:
I have built these projects:
master gecko-dev
master embedlite-components
branch 51673 sailfish-components-webview
master qtmozebed
next sailfish-browser
It works for me as it should...

**Comment by atatarov** [comment]
> > It works for me as it should...
> 
> Thanks for checking. I'll update all the relevant components again and give it another go

Now, I got it. WebView required this code
https://github.com/sailfishos/sailfish-browser/commit/1d99576b5d357c733dd4170c8c2a08a48e1e14f7#diff-7c0992668f1ab6f9e4c4af985ea53cc0b4b4a3059240a58c24173e8df248b951R390
I have been testing it for sailfish-browser, therefore its works for me

**Comment by rainemak** [comment]
If you need a common code for this, it should be going to https://github.com/sailfishos/sailfish-components-webview/blob/master/import/pickers/PickerOpener.qml . That already handles different file picking and ´<select>`.

**Comment by atatarov** [comment]
it is needed to wait for next prs:
https://git.sailfishos.org/mer-core/embedlite-components/merge_requests/118/diffs
https://github.com/sailfishos/sailfish-components-webview/pull/97 
https://github.com/sailfishos/sailfish-browser/pull/821

**Comment by llewelld** [comment]
With the others now merged, this should no longer be needed. Can it now be closed?

**Comment by atatarov** [comment]
> With the others now merged, this should no longer be needed. Can it now be closed?

It is needed for sailfish-browser --> contextMenu --> "Save Link" or "Save Image"

**Review by llewelld — COMMENTED** [review]
The code looks fine to me and works as expected. However currently there's no way for a developer to activate the "download menu" functionality of the webview as far as I can tell. One solution would be to expose the `popupOpener.downloadsEnabled` in WebView.qml, so that it can be set by developers.
```
    property alias downloadsEnabled: popupOpener.downloadsEnabled
```
Could something like this be added so that it's available?

**Review by llewelld — CHANGES_REQUESTED** [review]
Just now one pending request as far as I'm cconerned, to add the following to WebView.qml.
```
    property alias downloadsEnabled: popupOpener.downloadsEnabled
```

**Review by llewelld — APPROVED** [review]
Thanks @atatarov, this LGTM.

**Review by okodron — APPROVED** [review]

**Pull request merged**